### DISCUSSION TITLE
Dependency update workflow - can it be faster?

### DISCUSSION THREAD

**Discussion by camio** [microsoft/vcpkg#29847] [General]
Consider the case that someone is working on a project and sees the need to add an additional function to one of its dependencies. My understanding is that the vcpkg workflow for this would be as follows:

1. Fork the dependency library
2. Check out your fork
3. Make a new branch for it
4. Make changes
5. Commit
6. Push your branch to your fork
7. Update the top-level project's manifest to point to your fork
8. Run cmake configure
9. Re-open your IDE
10. Make changes to the top-level project to incorporate the changes
11. Build and test
12. Repeat steps 4-11 until you're happy with the change
13. Create a PR for your dependency
14. Wait until that gets incorporated to the main branch
15. Update the top-level project's manifest to point to the new dependency version
16. Create a PR for the top-level project.
17. Wait until that gets incorporated to the main branch.

The developer's feedback loop as they repeat steps 4-11 is quite long. Is there another way to do this?

Ideally, the workflow would be something like this:

1. Modify the manifest file so that it will use a local checkout of the dependency
2. Run cmake configure (this populates a local checkout of the dependency if it doesn't already exist)
3. Re-open your IDE (the dependency is now _part of_ the project)
4. Make changes to the dependency and top-level project as needed
5. Build and test
6. Repeat steps 4-5 until you're happy with the change
7. Create a PR for your dependency
8. Wait until that gets incorporated to the main branch
9. Update the top-level project's manifest to point to the new dependency version
10. Create a PR for the top-level project.
11. Wait until that gets incorporated to the main branch.

**Comment by JackBoosY** [comment]
You can add `--editable` to the end of the `install` command to ensure that the source file directory of the port will not be refreshed, and then commit and make patch until the verification is completed.

We have no solution for repeated modification after opening the PR: you always need CI to ensure the validity of the modification under all triplets.

**Comment by camio** [comment]
Thanks. Is there a way to pass command line arguments if one is using the recommended manifest mode?  When using --editable, how does one test the changes locally?

**Reply by JackBoosY** [reply]
`set(VCPKG_INSTALL_OPTIONS --editable)`.

When using `--editable`, vcpkg will check whether the source folder(_VCPKG_ROOT/buildtrees/<PORT>/src/<SOURCE_FOLDER>_) is exist. If not, vcpkg will extract the source archive.
If it exists, vcpkg will directly use the folder to configure / build instead of remove and re-extract it.
Therefore, you can make changes in the source folder for multiple times when your changes are not published.

**Reply by recsater** [reply]
> `set(VCPKG_INSTALL_OPTIONS --editable)`.
> 
> When using `--editable`, vcpkg will check whether the source folder(_VCPKG_ROOT/buildtrees//src/<SOURCE_FOLDER>_) is exist. If not, vcpkg will extract the source archive. If it exists, vcpkg will directly use the folder to configure / build instead of remove and re-extract it. Therefore, you can make changes in the source folder for multiple times when your changes are not published.

error: The option --editable is not supported in manifest mode.

**Reply by JackBoosY** [reply]
Can you please open a issue about adding `--editable` to manifest mode?

**Reply by recsater** [reply]
> Can you please open a issue about adding `--editable` to manifest mode?

What does it mean? Am I the only one having trouble with that option not working in manifest mode? May I ask why you're requesting to create an issue regarding this matter?

**Reply by JackBoosY** [reply]
It looks like option `--editable` is not allowed in manifest mode, but I think we need it.

**Reply by Osyotr** [reply]
@JackBoosY This option doesn't make sense in manifest mode.

**Reply by JackBoosY** [reply]
> @JackBoosY This option doesn't make sense in manifest mode.

Right I almost forgot this option implementation.

**Comment by Ikar-Rahl** [comment]
Hello, we are investigating the usability of vcpkg in our project and stumbled across this exact issue.

Often, we need to iterate over dependency source code, sometime, it is not even in order to publish a change, but to add a log somewhere, or dump an internal data to a file, typical debugging stuff.

The easiest way I have found is to have a port overlay and make patches for each change, but this is quite impractical to iterate rapidly.
